src/core/exec-invoke.c

```diff
@@ -4451,6 +4451,17 @@ int exec_invoke(
 
 #if ENABLE_UTMP
         if (context->utmp_id) {
+                _cleanup_free_ char *username_alloc = NULL;
+
+                if (!username && context->utmp_mode == EXEC_UTMP_USER) {
+                        username_alloc = uid_to_name(uid_is_valid(uid) ? uid : saved_uid);
+                        if (!username_alloc) {
+                                *exit_status = EXIT_USER;
+                                return log_oom();
+                        }
+                        username = username_alloc;
```

Comment:
No. `username` corresponds to the `User=` setting. This means that we assume User= is set in `build_environment` and possibly other places, breaking the semantics of e.g. `SetLoginEnvironment=`

Reply:
And `username` becomes dangling when username_alloc gets destroyed?

Reply:
> And `username` becomes dangling when username_alloc gets destroyed?

Indeed. Fixed.